### PR TITLE
#280

### PR DESCRIPTION
**Pull request by soma42**
see #280

### CODE REVIEW

File: src/Features/Blockcore.Features.MemoryPool/MempoolBehavior.cs [hunk 1]
```diff
@@ -162,6 +162,12 @@ private async Task OnMessageReceivedAsync(INetworkPeer peer, IncomingMessage mes
         {
             try
             {
+                if (this.initialBlockDownloadState.IsInitialBlockDownload())
```
Comment (dangershony) [inline src/Features/Blockcore.Features.MemoryPool/MempoolBehavior.cs:165]: ah interesting you want to completely invalidate mempool when in IBD

**Comment by dangershony** [comment]
As I see it the consequence of this change is that if a node expects funds and was offline for a few days, then the transaction that is sent to it when the nodes comes inline will not be seen at all until included in a block.

For example say someone wants to send me some coins now and I need to open my node and it will take say 10min to resync, if trx is sent inside then 10min then I will not see it as unconfirmed  

Two way to fix this:
- when IBD is finished ask for mempool trxs from a peer
- perhaps we use a different measure from what IBD uses (say maybe the limit should be further in the past)

**Comment by soma42** [comment]
@dangershony that is a very interesting point, and you are right, I won't see any mempool tx (pending for me or anybody else) until the point I'm synced. However, when I get a new additional connection, I might get it from the new peer.

I think I have looked earlier into this, because I wanted to understand the tx relay behavior because of mining.

Still, I think exactly as you describe it, is the correct behavior. Bitcoin is also not super eager with tx relay to not tress the network to much.

**Comment by soma42** [comment]
I think that was it: There were tx in the XDS mempool, and they did not get mined. So I started up a node, expecting to receive these tx, to look at them in the debugger, but they were not relayed to me. Then I was searching for what I should expect, and I think the result was, that that's by design. so there must be code somewhere, that knows when my connection started, and that ensures tx in mempool older than the connection date will not get relayed to me.

If I am right about that, it would also be correct to disable the mempool until IBD is finished, because I'm simply not supposed to see these tx by other parts of the protocol.

**Comment by dangershony** [comment]
>  so there must be code somewhere, that knows when my connection started, and that ensures tx in mempool older than the connection date will not get relayed to me.

I am not aware of such code, peers will just relay a trx upon receive (apart for the peer they received it form)

> I think the thing is that the mempool txs are mostly pushed by other peers, and I don't pull them (normally).

I there is a payload `MempoolPayload` that asks for all trx in the mempool frmo a peer.
I am not sure its used much though and how it behaves under load
Here is the method:

https://github.com/block-core/blockcore/blob/8595042e5edde22fee589df9e2711e994409b3c3/src/Features/Blockcore.Features.MemoryPool/MempoolBehavior.cs#L221

**Comment by dangershony** [comment]
@MithrilMan your thoughts?

**Comment by MithrilMan** [comment]
Placing that check in that method prevent to process any message related to memory pool and I'm not sure this is what we want (we need to look at it carefully)
basically it skips every payload managed by ProcessMessageAsync but during IBD. Now I don't remember if some of them may be used (e.g. by whitelisted peers)

I can't check on bitcoin source to compare the implementation, I'd like to know also how it handle transactions in mempool when the node shutdown (if it just discard them all keeping only our unconfirmed transaction generated by our wallet or if it stores everything)

Anyway that's a side question, the point here is if we want the memory pool to be network-unaware completly during IBD. Actually there isn't a reason on top of my head to process network messages during IBD

**Comment by dangershony** [comment]
> Now I don't remember if some of them may be used (e.g. by whitelisted peers)

Hmm that's possible that whitelist will allow it but there is not much point dusting IBD even for whitelisted peers because probably the trx will go in a block anyway.

> I can't check on bitcoin source to compare the implementation, I'd like to know also how it handle transactions in mempool when the node shutdown

Perhaps its a good idea to check core, @soma42 what is dogecoin doing? (they are up to date with core right?)

**Review by dangershony — APPROVED** [review]

**Review by sondreb — APPROVED** [review]

**Pull request merged**